tools/halide_image.h

```diff
@@ -1,278 +1,696 @@
-// This header defines a simple Image class which wraps a buffer_t. This is
-// useful when interacting with a statically-compiled Halide pipeline emitted by
-// Func::compile_to_file, when you do not want to link your processing program
-// against Halide.h/libHalide.a.
+// A simple Image class which wraps a buffer_t. This is useful
+// when interacting with a statically-compiled Halide pipeline emitted
+// by Func::compile_to_file, when you do not want to link your
+// processing program against Halide.h/libHalide.a.
 
-#ifndef HALIDE_TOOLS_IMAGE_H
-#define HALIDE_TOOLS_IMAGE_H
+#ifndef HALIDE_RUNTIME_IMAGE_H
+#define HALIDE_RUNTIME_IMAGE_H
 
-#include <cassert>
-#include <cstdlib>
-#include <limits>
 #include <memory>
 #include <vector>
-#include <stdint.h>  // <cstdint> requires C++11
+#include <cassert>
+#include <stdint.h>
+#include <string.h>
 
 #include "HalideRuntime.h"
 
 namespace Halide {
 namespace Tools {
 
-template<typename T>
-class Image {
-    struct Contents {
-        Contents(const buffer_t &b, uint8_t *a) : buf(b), ref_count(1), alloc(a) {}
-        buffer_t buf;
-        int ref_count;
-        uint8_t *alloc;
+template<typename Fn>
+void for_each_element(const buffer_t &buf, Fn f);
```

Comment:
For now I deleted those because I think they were needlessly clever for the
uses that exist. I'll resurrect if they're the right thing.

On Thu, Jul 28, 2016 at 2:16 PM, Zalman Stern <notifications@github.com>
wrote:

> In tools/halide_image.h
> <https://github.com/halide/Halide/pull/1402#discussion_r72704378>:
>
> >
> >  #include "HalideRuntime.h"
> >
> >  namespace Halide {
> >  namespace Tools {
> >
> > -template<typename T>
> > -class Image {
> > -    struct Contents {
> > -        Contents(const buffer_t &b, uint8_t *a) : buf(b), ref_count(1), alloc(a) {}
> > -        buffer_t buf;
> > -        int ref_count;
> > -        uint8_t *alloc;
> > +template<typename Fn>
> > +void for_each_element(const buffer_t &buf, Fn f);
>
> Also, did you decide not to do the version which calls a function on
> contiguous runs of memory?
>
> —
> You are receiving this because you authored the thread.
> Reply to this email directly, view it on GitHub
> <https://github.com/halide/Halide/pull/1402/files/44f3bd8358b400618e1a033faddb06a39fe1387d#r72704378>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AAfdRsrEIIGfMUuf4AD_vjGmLXsAjukeks5qaRwYgaJpZM4JTMHs>
> .
>